crates/ruff_server/src/server.rs

```diff
@@ -146,11 +97,14 @@ impl Server {
     #[allow(clippy::needless_pass_by_value)] // this is because we aren't using `next_request_id` yet.
     fn event_loop(
         connection: &Connection,
-        session: Session,
+        client_capabilities: &ClientCapabilities,
+        mut session: Session,
         worker_threads: NonZeroUsize,
-        _next_request_id: AtomicI32,
     ) -> crate::Result<()> {
-        let mut scheduler = schedule::Scheduler::new(session, worker_threads, &connection.sender);
+        let mut scheduler =
+            schedule::Scheduler::new(&mut session, worker_threads, &connection.sender);
+
+        Self::try_register_capabilities(client_capabilities, &mut scheduler);
```

Comment:
Is the right way to think of this that:

- We might have some responses on `connection.receiver` already (but they can be buffered).
- We kick off a "blocking" request to register capabilities, and wait for the client to respond.
- Once the client responds, we move to to processing messages on `connection.receiver`.

Reply:
Ahh ok makes sense. Is it important that the dynamic capabilities response is handled "first", or it's ok if they're out-of-order? (Do we enforce the order in any way?)

Reply:
> (Do we enforce the order in any way?)

We don't enforce the order - we rely on `lsp_server::Connection` to give us messages in the order the client sent them. The LSP specification requires that [requests on both sides are processed in the order that they're received](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#messageOrdering), but it doesn't say that all incoming requests need to be handled before responses, or vice-versa.

> Is it important that the dynamic capabilities response is handled "first", or it's ok if they're out-of-order?

In this specific instance, it's absolutely fine to have this happen out-of-order because the response handler doesn't do anything besides log a success. Technically, the client could wait an arbitrary amount before sending this response if it wanted to (though that would be rude).

I think this scenario would also work in the general case:
```
1. client sends request A to server
2. server sends request B to client
3. server sends response A to client
4. client sends response B response to server
```

The response handler for `response B` would be run with a potentially different state then when `request B` was sent occurred if `request A` was a local task that modified the state, but the response handler shouldn't have any guarantees on execution order to begin with.